src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/VariableBinder.cs

```diff
@@ -31,6 +32,27 @@ private VariableBinder()
         {
         }
 
+        public override Expression Visit(Expression node)
+        {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we use the StackGuard utility here to probe for sufficient stackand continue the work on
+            // another thread when we run out of stack space.
+            if (!_guard.TryEnterOnCurrentStack())
+            {
+                return _guard.RunOnEmptyStack(s =>
+                {
+                    var t = (Tuple<VariableBinder, Expression>)s;
+                    return t.Item1.Visit(t.Item2);
```

Comment:
If RunOnEmptyStack is generic on the type of s, perhaps we would not need a cast here?